<!DOCTYPE html><html><head><meta charset="utf-8"><style>@font-face {
  font-family: octicons-anchor;
  src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAYcAA0AAAAACjQAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABMAAAABwAAAAca8vGTk9TLzIAAAFMAAAARAAAAFZG1VHVY21hcAAAAZAAAAA+AAABQgAP9AdjdnQgAAAB0AAAAAQAAAAEACICiGdhc3AAAAHUAAAACAAAAAj//wADZ2x5ZgAAAdwAAADRAAABEKyikaNoZWFkAAACsAAAAC0AAAA2AtXoA2hoZWEAAALgAAAAHAAAACQHngNFaG10eAAAAvwAAAAQAAAAEAwAACJsb2NhAAADDAAAAAoAAAAKALIAVG1heHAAAAMYAAAAHwAAACABEAB2bmFtZQAAAzgAAALBAAAFu3I9x/Nwb3N0AAAF/AAAAB0AAAAvaoFvbwAAAAEAAAAAzBdyYwAAAADP2IQvAAAAAM/bz7t4nGNgZGFgnMDAysDB1Ml0hoGBoR9CM75mMGLkYGBgYmBlZsAKAtJcUxgcPsR8iGF2+O/AEMPsznAYKMwIkgMA5REMOXicY2BgYGaAYBkGRgYQsAHyGMF8FgYFIM0ChED+h5j//yEk/3KoSgZGNgYYk4GRCUgwMaACRoZhDwCs7QgGAAAAIgKIAAAAAf//AAJ4nHWMMQrCQBBF/0zWrCCIKUQsTDCL2EXMohYGSSmorScInsRGL2DOYJe0Ntp7BK+gJ1BxF1stZvjz/v8DRghQzEc4kIgKwiAppcA9LtzKLSkdNhKFY3HF4lK69ExKslx7Xa+vPRVS43G98vG1DnkDMIBUgFN0MDXflU8tbaZOUkXUH0+U27RoRpOIyCKjbMCVejwypzJJG4jIwb43rfl6wbwanocrJm9XFYfskuVC5K/TPyczNU7b84CXcbxks1Un6H6tLH9vf2LRnn8Ax7A5WQAAAHicY2BkYGAA4teL1+yI57f5ysDNwgAC529f0kOmWRiYVgEpDgYmEA8AUzEKsQAAAHicY2BkYGB2+O/AEMPCAAJAkpEBFbAAADgKAe0EAAAiAAAAAAQAAAAEAAAAAAAAKgAqACoAiAAAeJxjYGRgYGBhsGFgYgABEMkFhAwM/xn0QAIAD6YBhwB4nI1Ty07cMBS9QwKlQapQW3VXySvEqDCZGbGaHULiIQ1FKgjWMxknMfLEke2A+IJu+wntrt/QbVf9gG75jK577Lg8K1qQPCfnnnt8fX1NRC/pmjrk/zprC+8D7tBy9DHgBXoWfQ44Av8t4Bj4Z8CLtBL9CniJluPXASf0Lm4CXqFX8Q84dOLnMB17N4c7tBo1AS/Qi+hTwBH4rwHHwN8DXqQ30XXAS7QaLwSc0Gn8NuAVWou/gFmnjLrEaEh9GmDdDGgL3B4JsrRPDU2hTOiMSuJUIdKQQayiAth69r6akSSFqIJuA19TrzCIaY8sIoxyrNIrL//pw7A2iMygkX5vDj+G+kuoLdX4GlGK/8Lnlz6/h9MpmoO9rafrz7ILXEHHaAx95s9lsI7AHNMBWEZHULnfAXwG9/ZqdzLI08iuwRloXE8kfhXYAvE23+23DU3t626rbs8/8adv+9DWknsHp3E17oCf+Z48rvEQNZ78paYM38qfk3v/u3l3u3GXN2Dmvmvpf1Srwk3pB/VSsp512bA/GG5i2WJ7wu430yQ5K3nFGiOqgtmSB5pJVSizwaacmUZzZhXLlZTq8qGGFY2YcSkqbth6aW1tRmlaCFs2016m5qn36SbJrqosG4uMV4aP2PHBmB3tjtmgN2izkGQyLWprekbIntJFing32a5rKWCN/SdSoga45EJykyQ7asZvHQ8PTm6cslIpwyeyjbVltNikc2HTR7YKh9LBl9DADC0U/jLcBZDKrMhUBfQBvXRzLtFtjU9eNHKin0x5InTqb8lNpfKv1s1xHzTXRqgKzek/mb7nB8RZTCDhGEX3kK/8Q75AmUM/eLkfA+0Hi908Kx4eNsMgudg5GLdRD7a84npi+YxNr5i5KIbW5izXas7cHXIMAau1OueZhfj+cOcP3P8MNIWLyYOBuxL6DRylJ4cAAAB4nGNgYoAALjDJyIAOWMCiTIxMLDmZedkABtIBygAAAA==) format('woff');
}

* {
    box-sizing: border-box;
}

body {
    width: 980px;
    margin-right: auto;
    margin-left: auto;
}

body .markdown-body {
    padding: 45px;
    border: 1px solid #ddd;
    border-radius: 3px;
    word-wrap: break-word;
}

pre {
    font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}

.markdown-body {
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  color: #333;
  font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 16px;
  line-height: 1.6;
  word-wrap: break-word;
}

.markdown-body a {
  background-color: transparent;
}

.markdown-body a:active,
.markdown-body a:hover {
  outline: 0;
}

.markdown-body strong {
  font-weight: bold;
}

.markdown-body h1 {
  font-size: 2em;
  margin: 0.67em 0;
}

.markdown-body img {
  border: 0;
}

.markdown-body hr {
  box-sizing: content-box;
  height: 0;
}

.markdown-body pre {
  overflow: auto;
}

.markdown-body code,
.markdown-body kbd,
.markdown-body pre {
  font-family: monospace, monospace;
  font-size: 1em;
}

.markdown-body input {
  color: inherit;
  font: inherit;
  margin: 0;
}

.markdown-body html input[disabled] {
  cursor: default;
}

.markdown-body input {
  line-height: normal;
}

.markdown-body input[type="checkbox"] {
  box-sizing: border-box;
  padding: 0;
}

.markdown-body table {
  border-collapse: collapse;
  border-spacing: 0;
}

.markdown-body td,
.markdown-body th {
  padding: 0;
}

.markdown-body input {
  font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}

.markdown-body a {
  color: #4078c0;
  text-decoration: none;
}

.markdown-body a:hover,
.markdown-body a:active {
  text-decoration: underline;
}

.markdown-body hr {
  height: 0;
  margin: 15px 0;
  overflow: hidden;
  background: transparent;
  border: 0;
  border-bottom: 1px solid #ddd;
}

.markdown-body hr:before {
  display: table;
  content: "";
}

.markdown-body hr:after {
  display: table;
  clear: both;
  content: "";
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-top: 15px;
  margin-bottom: 15px;
  line-height: 1.1;
}

.markdown-body h1 {
  font-size: 30px;
}

.markdown-body h2 {
  font-size: 21px;
}

.markdown-body h3 {
  font-size: 16px;
}

.markdown-body h4 {
  font-size: 14px;
}

.markdown-body h5 {
  font-size: 12px;
}

.markdown-body h6 {
  font-size: 11px;
}

.markdown-body blockquote {
  margin: 0;
}

.markdown-body ul,
.markdown-body ol {
  padding: 0;
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body ol ol,
.markdown-body ul ol {
  list-style-type: lower-roman;
}

.markdown-body ul ul ol,
.markdown-body ul ol ol,
.markdown-body ol ul ol,
.markdown-body ol ol ol {
  list-style-type: lower-alpha;
}

.markdown-body dd {
  margin-left: 0;
}

.markdown-body code {
  font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
  font-size: 12px;
}

.markdown-body pre {
  margin-top: 0;
  margin-bottom: 0;
  font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}

.markdown-body .select::-ms-expand {
  opacity: 0;
}

.markdown-body .octicon {
  font: normal normal normal 16px/1 octicons-anchor;
  display: inline-block;
  text-decoration: none;
  text-rendering: auto;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.markdown-body .octicon-link:before {
  content: '\f05c';
}

.markdown-body:before {
  display: table;
  content: "";
}

.markdown-body:after {
  display: table;
  clear: both;
  content: "";
}

.markdown-body>*:first-child {
  margin-top: 0 !important;
}

.markdown-body>*:last-child {
  margin-bottom: 0 !important;
}

.markdown-body a:not([href]) {
  color: inherit;
  text-decoration: none;
}

.markdown-body .anchor {
  display: inline-block;
  padding-right: 2px;
  margin-left: -18px;
}

.markdown-body .anchor:focus {
  outline: none;
}

.markdown-body h1,
.markdown-body h2,
.markdown-body h3,
.markdown-body h4,
.markdown-body h5,
.markdown-body h6 {
  margin-top: 1em;
  margin-bottom: 16px;
  font-weight: bold;
  line-height: 1.4;
}

.markdown-body h1 .octicon-link,
.markdown-body h2 .octicon-link,
.markdown-body h3 .octicon-link,
.markdown-body h4 .octicon-link,
.markdown-body h5 .octicon-link,
.markdown-body h6 .octicon-link {
  color: #000;
  vertical-align: middle;
  visibility: hidden;
}

.markdown-body h1:hover .anchor,
.markdown-body h2:hover .anchor,
.markdown-body h3:hover .anchor,
.markdown-body h4:hover .anchor,
.markdown-body h5:hover .anchor,
.markdown-body h6:hover .anchor {
  text-decoration: none;
}

.markdown-body h1:hover .anchor .octicon-link,
.markdown-body h2:hover .anchor .octicon-link,
.markdown-body h3:hover .anchor .octicon-link,
.markdown-body h4:hover .anchor .octicon-link,
.markdown-body h5:hover .anchor .octicon-link,
.markdown-body h6:hover .anchor .octicon-link {
  visibility: visible;
}

.markdown-body h1 {
  padding-bottom: 0.3em;
  font-size: 2.25em;
  line-height: 1.2;
  border-bottom: 1px solid #eee;
}

.markdown-body h1 .anchor {
  line-height: 1;
}

.markdown-body h2 {
  padding-bottom: 0.3em;
  font-size: 1.75em;
  line-height: 1.225;
  border-bottom: 1px solid #eee;
}

.markdown-body h2 .anchor {
  line-height: 1;
}

.markdown-body h3 {
  font-size: 1.5em;
  line-height: 1.43;
}

.markdown-body h3 .anchor {
  line-height: 1.2;
}

.markdown-body h4 {
  font-size: 1.25em;
}

.markdown-body h4 .anchor {
  line-height: 1.2;
}

.markdown-body h5 {
  font-size: 1em;
}

.markdown-body h5 .anchor {
  line-height: 1.1;
}

.markdown-body h6 {
  font-size: 1em;
  color: #777;
}

.markdown-body h6 .anchor {
  line-height: 1.1;
}

.markdown-body p,
.markdown-body blockquote,
.markdown-body ul,
.markdown-body ol,
.markdown-body dl,
.markdown-body table,
.markdown-body pre {
  margin-top: 0;
  margin-bottom: 16px;
}

.markdown-body hr {
  height: 4px;
  padding: 0;
  margin: 16px 0;
  background-color: #e7e7e7;
  border: 0 none;
}

.markdown-body ul,
.markdown-body ol {
  padding-left: 2em;
}

.markdown-body ul ul,
.markdown-body ul ol,
.markdown-body ol ol,
.markdown-body ol ul {
  margin-top: 0;
  margin-bottom: 0;
}

.markdown-body li>p {
  margin-top: 16px;
}

.markdown-body dl {
  padding: 0;
}

.markdown-body dl dt {
  padding: 0;
  margin-top: 16px;
  font-size: 1em;
  font-style: italic;
  font-weight: bold;
}

.markdown-body dl dd {
  padding: 0 16px;
  margin-bottom: 16px;
}

.markdown-body blockquote {
  padding: 0 15px;
  color: #777;
  border-left: 4px solid #ddd;
}

.markdown-body blockquote>:first-child {
  margin-top: 0;
}

.markdown-body blockquote>:last-child {
  margin-bottom: 0;
}

.markdown-body table {
  display: block;
  width: 100%;
  overflow: auto;
  word-break: normal;
  word-break: keep-all;
}

.markdown-body table th {
  font-weight: bold;
}

.markdown-body table th,
.markdown-body table td {
  padding: 6px 13px;
  border: 1px solid #ddd;
}

.markdown-body table tr {
  background-color: #fff;
  border-top: 1px solid #ccc;
}

.markdown-body table tr:nth-child(2n) {
  background-color: #f8f8f8;
}

.markdown-body img {
  max-width: 100%;
  box-sizing: content-box;
  background-color: #fff;
}

.markdown-body code {
  padding: 0;
  padding-top: 0.2em;
  padding-bottom: 0.2em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(0,0,0,0.04);
  border-radius: 3px;
}

.markdown-body code:before,
.markdown-body code:after {
  letter-spacing: -0.2em;
  content: "\00a0";
}

.markdown-body pre>code {
  padding: 0;
  margin: 0;
  font-size: 100%;
  word-break: normal;
  white-space: pre;
  background: transparent;
  border: 0;
}

.markdown-body .highlight {
  margin-bottom: 16px;
}

.markdown-body .highlight pre,
.markdown-body pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f7f7f7;
  border-radius: 3px;
}

.markdown-body .highlight pre {
  margin-bottom: 0;
  word-break: normal;
}

.markdown-body pre {
  word-wrap: normal;
}

.markdown-body pre code {
  display: inline;
  max-width: initial;
  padding: 0;
  margin: 0;
  overflow: initial;
  line-height: inherit;
  word-wrap: normal;
  background-color: transparent;
  border: 0;
}

.markdown-body pre code:before,
.markdown-body pre code:after {
  content: normal;
}

.markdown-body kbd {
  display: inline-block;
  padding: 3px 5px;
  font-size: 11px;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb;
}

.markdown-body .pl-c {
  color: #969896;
}

.markdown-body .pl-c1,
.markdown-body .pl-s .pl-v {
  color: #0086b3;
}

.markdown-body .pl-e,
.markdown-body .pl-en {
  color: #795da3;
}

.markdown-body .pl-s .pl-s1,
.markdown-body .pl-smi {
  color: #333;
}

.markdown-body .pl-ent {
  color: #63a35c;
}

.markdown-body .pl-k {
  color: #a71d5d;
}

.markdown-body .pl-pds,
.markdown-body .pl-s,
.markdown-body .pl-s .pl-pse .pl-s1,
.markdown-body .pl-sr,
.markdown-body .pl-sr .pl-cce,
.markdown-body .pl-sr .pl-sra,
.markdown-body .pl-sr .pl-sre {
  color: #183691;
}

.markdown-body .pl-v {
  color: #ed6a43;
}

.markdown-body .pl-id {
  color: #b52a1d;
}

.markdown-body .pl-ii {
  background-color: #b52a1d;
  color: #f8f8f8;
}

.markdown-body .pl-sr .pl-cce {
  color: #63a35c;
  font-weight: bold;
}

.markdown-body .pl-ml {
  color: #693a17;
}

.markdown-body .pl-mh,
.markdown-body .pl-mh .pl-en,
.markdown-body .pl-ms {
  color: #1d3e81;
  font-weight: bold;
}

.markdown-body .pl-mq {
  color: #008080;
}

.markdown-body .pl-mi {
  color: #333;
  font-style: italic;
}

.markdown-body .pl-mb {
  color: #333;
  font-weight: bold;
}

.markdown-body .pl-md {
  background-color: #ffecec;
  color: #bd2c00;
}

.markdown-body .pl-mi1 {
  background-color: #eaffea;
  color: #55a532;
}

.markdown-body .pl-mdr {
  color: #795da3;
  font-weight: bold;
}

.markdown-body .pl-mo {
  color: #1d3e81;
}

.markdown-body kbd {
  display: inline-block;
  padding: 3px 5px;
  font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb;
}

.markdown-body .plan-price-unit {
  color: #767676;
  font-weight: normal;
}

.markdown-body .task-list-item {
  list-style-type: none;
}

.markdown-body .task-list-item+.task-list-item {
  margin-top: 3px;
}

.markdown-body .task-list-item input {
  margin: 0 0.35em 0.25em -1.6em;
  vertical-align: middle;
}

.markdown-body .plan-choice {
  padding: 15px;
  padding-left: 40px;
  display: block;
  border: 1px solid #e0e0e0;
  position: relative;
  font-weight: normal;
  background-color: #fafafa;
}

.markdown-body .plan-choice.open {
  background-color: #fff;
}

.markdown-body .plan-choice.open .plan-choice-seat-breakdown {
  display: block;
}

.markdown-body .plan-choice-free {
  border-radius: 3px 3px 0 0;
}

.markdown-body .plan-choice-paid {
  border-radius: 0 0 3px 3px;
  border-top: 0;
  margin-bottom: 20px;
}

.markdown-body .plan-choice-radio {
  position: absolute;
  left: 15px;
  top: 18px;
}

.markdown-body .plan-choice-exp {
  color: #999;
  font-size: 12px;
  margin-top: 5px;
}

.markdown-body .plan-choice-seat-breakdown {
  margin-top: 10px;
  display: none;
}

.markdown-body :checked+.radio-label {
  z-index: 1;
  position: relative;
  border-color: #4078c0;
}
</style><title>README</title></head><body><article class="markdown-body"><h1>
<a id="user-content-mtree" class="anchor" href="#mtree" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>MTREE</h1>

<pre><code>A program to predict interactions using mirror trees.
        Created by Sergio Castillo, Joan Martí &amp; Adrià Pérez
</code></pre>

<h2>
<a id="user-content-install-mtree" class="anchor" href="#install-mtree" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Install mtree</h2>

<blockquote>
<p>Move to the root directory of the program and run these commands:</p>
</blockquote>

<div class="highlight highlight-source-shell"><pre>sudo python3 setup.py install</pre></div>

<h2>
<a id="user-content-usage-of-mtree" class="anchor" href="#usage-of-mtree" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Usage of mtree</h2>

<div class="highlight highlight-source-shell"><pre>mtree -i input.fa -db uniprot.fa --data datafolder -g graph.html -c 0.8 -sp 8 <span class="pl-k">&gt;</span> prediction.out</pre></div>

<h3>
<a id="user-content-options" class="anchor" href="#options" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Options</h3>

<ul>
<li>
<p><strong>-h,    --help</strong>  </p>

<pre><code>Shows the options of the program.
</code></pre>
</li>
<li>
<p><strong>-i,    --input</strong>      FILENAME</p>

<pre><code>REQUIRED. Input FASTA with the sequences to predict interactions.
</code></pre>
</li>
<li>
<p><strong>-o,    --output</strong>      FILENAME</p>

<pre><code>OPTIONAL. Filename of the output. If it's not defined mtree will print to STDOUT.
</code></pre>
</li>
<li>
<p><strong>-db,   --database</strong>      FILENAME</p>

<pre><code>REQUIRED. Sequence DATABASE in FASTA format. Will be used to find homologs using jackhmmer.
</code></pre>
</li>
<li>
<p><strong>-d,    --data</strong>      DIRECTORY</p>

<pre><code>REQUIRED. Directory with the data files used by mtree.
</code></pre>
</li>
<li>
<p><strong>-sp,   --species</strong>      INT</p>

<pre><code>OPTIONAL. Number of species needed to build the phylogenetic trees. Default = 10.
</code></pre>
</li>
<li>
<p><strong>-c,    --cutoff</strong>      FLOAT</p>

<pre><code>OPTIONAL. Pearson correlation cutoff to predict interactions. Default = 0.7.
</code></pre>
</li>
<li>
<p><strong>-g,    --graph</strong>     FILENAME</p>

<pre><code>OPTIONAL. If this option is used, mtree will create an html file with the graph drawn using cytoscape.js.
</code></pre>
</li>
<li>
<p><strong>-ints, --ints</strong>     FILENAME</p>

<pre><code>OPTIONAL. USED ONLY FOR TESTING/DEBUGGING! Considers only interactions in FILENAME.
</code></pre>
</li>
</ul>

<h3>
<a id="user-content-output-of-the-program" class="anchor" href="#output-of-the-program" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Output of the program</h3>

<blockquote>
<p>The program outputs a tabular file to the specified path (using the option -o). This file is of the following form:</p>
</blockquote>

<table>
<thead>
<tr>
<th align="center">SEQ1</th>
<th align="center">SEQ2</th>
<th align="center">Pearson</th>
<th align="center">Spearman</th>
<th align="center">r_Adjusted</th>
<th align="center">Type</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">CLOCK</td>
<td align="center">BMAL</td>
<td align="center">0.7</td>
<td align="center">0.6</td>
<td align="center">0.8</td>
<td align="center">Int</td>
</tr>
<tr>
<td align="center">EGFR</td>
<td align="center">TNFA</td>
<td align="center">0.3</td>
<td align="center">0.2</td>
<td align="center">0.4</td>
<td align="center">NonInt</td>
</tr>
</tbody>
</table>

<h1>
<a id="user-content-testing-the-program" class="anchor" href="#testing-the-program" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Testing the program</h1>

<h2>
<a id="user-content-downloading-the-databases" class="anchor" href="#downloading-the-databases" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Downloading the databases</h2>

<blockquote>
<p><strong>mtree</strong> needs a database of FASTA sequences to work. We used Swiss-Prot.
Also, for training purposes, we used INTACT and NEGATOME. These two
are NOT necessary for the program to work, but we used them to evaluate our method.</p>
</blockquote>

<h3>
<a id="user-content-create-directory" class="anchor" href="#create-directory" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Create directory</h3>

<div class="highlight highlight-source-shell"><pre>mkdir db</pre></div>

<h3>
<a id="user-content-download-uniprot" class="anchor" href="#download-uniprot" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Download uniprot</h3>

<div class="highlight highlight-source-shell"><pre>wget ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz \
     -O db/uniprot_sprot.dat.gz</pre></div>

<h3>
<a id="user-content-subsetting-the-database" class="anchor" href="#subsetting-the-database" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Subsetting the database</h3>

<blockquote>
<p>Because we are always working with animals we filtered the database. This is NOT required.</p>
</blockquote>

<h4>
<a id="user-content-downloading-the-animals-list" class="anchor" href="#downloading-the-animals-list" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Downloading the animals list</h4>

<blockquote>
<p>Go to the website and download all the animal species names.
Save it as animals.txt
This step is NOT necessary. It was only useful for testing.</p>
</blockquote>

<ul>
<li><a href="http://www.ncbi.nlm.nih.gov/genome/browse/">http://www.ncbi.nlm.nih.gov/genome/browse/</a></li>
</ul>

<h4>
<a id="user-content-filter-the-fasta" class="anchor" href="#filter-the-fasta" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Filter the FASTA</h4>

<div class="highlight highlight-source-shell"><pre>zcat db/uniprot_sprot.dat.gz <span class="pl-k">|</span> \
perl -e <span class="pl-s"><span class="pl-pds">'</span></span>
<span class="pl-s">%specs = ();</span>
<span class="pl-s">open $fh, "&lt;animals.txt ";</span>
<span class="pl-s">while(&lt;$fh&gt;) {</span>
<span class="pl-s">    chomp;</span>
<span class="pl-s">    $specs{$_} = 1;</span>
<span class="pl-s">};</span>
<span class="pl-s">local $/ = "&gt;";</span>
<span class="pl-s">while(&lt;&gt;){</span>
<span class="pl-s">    chomp;</span>
<span class="pl-s">    ($name, @seq) = split /\n/;</span>
<span class="pl-s">    $sp = "";</span>
<span class="pl-s">    if ($name =~ m/OS=(.*?) GN/) {</span>
<span class="pl-s">        $sp = $1;</span>
<span class="pl-s">    } else {</span>
<span class="pl-s">        next</span>
<span class="pl-s">    };</span>
<span class="pl-s">    if (not exists $specs{$sp}){</span>
<span class="pl-s">        next;</span>
<span class="pl-s">    }  </span>
<span class="pl-s">    print "&gt;$name\n", join("\n", @seq), "\n";</span>
<span class="pl-s">} <span class="pl-pds">'</span></span> <span class="pl-k">&gt;</span> db/animals_uniprot.fa</pre></div>

<h3>
<a id="user-content-create-testing-sets" class="anchor" href="#create-testing-sets" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Create testing sets</h3>

<blockquote>
<p>We needed a set of interacting proteins and a set of non-interacting proteins.
We downloaded the former from INTACT, and the latter from Negatome</p>
</blockquote>

<h4>
<a id="user-content-download-intact" class="anchor" href="#download-intact" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Download INTACT</h4>

<div class="highlight highlight-source-shell"><pre>wget ftp://ftp.ebi.ac.uk/pub/databases/intact/current/psimitab/intact.txt -O db/intact.tbl</pre></div>

<h4>
<a id="user-content-download-negatome" class="anchor" href="#download-negatome" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Download NEGATOME</h4>

<div class="highlight highlight-source-shell"><pre>wget http://mips.helmholtz-muenchen.de/proj/ppi/negatome/manual_stringent.txt -O db/negatome.tbl

<span class="pl-c"># Get some FASTAs from negatome</span>
python3 bin/create_sets.py -i db/NonInt.tbl -db db/animals_uniprot.fa -o 300_negatome.fa --num 300</pre></div>

<h4>
<a id="user-content-run-mtree-with-positives" class="anchor" href="#run-mtree-with-positives" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Run mtree with positives</h4>

<div class="highlight highlight-source-shell"><pre>mtree -i 300_negatome.fa -db db/animals_uniprot.fa -ints db/intact.tbl --data data/ -sp 10 <span class="pl-k">|</span> \
    perl -e <span class="pl-s"><span class="pl-pds">'</span></span>
<span class="pl-s">    print &lt;&gt;;</span>
<span class="pl-s">    while(&lt;&gt;){</span>
<span class="pl-s">        chomp;</span>
<span class="pl-s">        @cols = split /\t/;</span>
<span class="pl-s">        print "$cols[0]\t$cols[1]\t$cols[2]\t$cols[3]\t$cols[4]\tYES\n";</span>
<span class="pl-s">    }<span class="pl-pds">'</span></span> <span class="pl-k">&gt;</span> testing.tbl</pre></div>

<h4>
<a id="user-content-run-mtree-with-negatives" class="anchor" href="#run-mtree-with-negatives" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Run mtree with negatives</h4>

<div class="highlight highlight-source-shell"><pre>mtree -i 300_negatome.fa -db db/animals_uniprot.fa -ints db/NonInt.tbl --data data/ -sp 10 <span class="pl-k">|</span> \
    perl -e <span class="pl-s"><span class="pl-pds">'</span></span>
<span class="pl-s">    print &lt;&gt;;</span>
<span class="pl-s">    while(&lt;&gt;){</span>
<span class="pl-s">        chomp;</span>
<span class="pl-s">        @cols = split /\t/;</span>
<span class="pl-s">        print "$cols[0]\t$cols[1]\t$cols[2]\t$cols[3]\t$cols[4]\tNO\n";</span>
<span class="pl-s">    }<span class="pl-pds">'</span></span> <span class="pl-k">&gt;&gt;</span> testing.tbl</pre></div>

<h2>
<a id="user-content-plots---r-console" class="anchor" href="#plots---r-console" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Plots - R console</h2>

<div class="highlight highlight-source-r"><pre>library(<span class="pl-smi">ggplot2</span>)
<span class="pl-smi">data</span> <span class="pl-k">&lt;-</span>read.table(<span class="pl-v">file</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>testing.tbl<span class="pl-pds">"</span></span>)
ggplot(<span class="pl-smi">data</span>) <span class="pl-k">+</span>
    geom_point(aes(<span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-smi">Type</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-smi">Pearson</span>, <span class="pl-v">color</span><span class="pl-k">=</span><span class="pl-smi">Type</span>), <span class="pl-v">position</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>jitter<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    xlab(<span class="pl-s"><span class="pl-pds">"</span><span class="pl-pds">"</span></span>) <span class="pl-k">+</span> ylab(<span class="pl-s"><span class="pl-pds">"</span>Coevolution (Pearson)<span class="pl-cce">\n</span><span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    theme_bw() <span class="pl-k">+</span>
    geom_hline(<span class="pl-v">yintercept</span><span class="pl-k">=</span><span class="pl-c1">0.7</span>, <span class="pl-v">linetype</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>dashed<span class="pl-pds">"</span></span>, <span class="pl-v">alpha</span><span class="pl-k">=</span><span class="pl-c1">0.6</span>) <span class="pl-k">+</span>
    geom_hline(<span class="pl-v">yintercept</span><span class="pl-k">=</span><span class="pl-c1">0.8</span>, <span class="pl-v">linetype</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>dashed<span class="pl-pds">"</span></span>, <span class="pl-v">alpha</span><span class="pl-k">=</span><span class="pl-c1">0.6</span>) <span class="pl-k">+</span>
    geom_hline(<span class="pl-v">yintercept</span><span class="pl-k">=</span><span class="pl-c1">0.9</span>, <span class="pl-v">linetype</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>dashed<span class="pl-pds">"</span></span>, <span class="pl-v">alpha</span><span class="pl-k">=</span><span class="pl-c1">0.6</span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">0.7</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.72</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>P = 0.78<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">0.7</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.82</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>P = 0.72<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">0.7</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.92</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>P = 0.63<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">2.4</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.72</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>R = 0.12<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">2.4</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.82</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>R = 0.07<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">2.4</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.92</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>R = 0.03<span class="pl-pds">"</span></span>)

ggplot(<span class="pl-smi">data</span>) <span class="pl-k">+</span>
    geom_point(aes(<span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-smi">Type</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-smi">Spearman</span>, <span class="pl-v">color</span><span class="pl-k">=</span><span class="pl-smi">Type</span>), <span class="pl-v">position</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>jitter<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    xlab(<span class="pl-s"><span class="pl-pds">"</span><span class="pl-pds">"</span></span>) <span class="pl-k">+</span> ylab(<span class="pl-s"><span class="pl-pds">"</span>Coevolution (Spearman)<span class="pl-cce">\n</span><span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    theme_bw() <span class="pl-k">+</span>
    geom_hline(<span class="pl-v">yintercept</span><span class="pl-k">=</span><span class="pl-c1">0.7</span>, <span class="pl-v">linetype</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>dashed<span class="pl-pds">"</span></span>, <span class="pl-v">alpha</span><span class="pl-k">=</span><span class="pl-c1">0.6</span>) <span class="pl-k">+</span>
    geom_hline(<span class="pl-v">yintercept</span><span class="pl-k">=</span><span class="pl-c1">0.8</span>, <span class="pl-v">linetype</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>dashed<span class="pl-pds">"</span></span>, <span class="pl-v">alpha</span><span class="pl-k">=</span><span class="pl-c1">0.6</span>) <span class="pl-k">+</span>
    geom_hline(<span class="pl-v">yintercept</span><span class="pl-k">=</span><span class="pl-c1">0.9</span>, <span class="pl-v">linetype</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>dashed<span class="pl-pds">"</span></span>, <span class="pl-v">alpha</span><span class="pl-k">=</span><span class="pl-c1">0.6</span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">0.7</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.72</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>P = 0,77<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">0.7</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.82</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>P = 0,80<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">0.7</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.92</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>P = 0.71<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">2.4</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.72</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>R = 0.08<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">2.4</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.82</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>R = 0.06<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">2.4</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.92</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>R = 0.02<span class="pl-pds">"</span></span>)

ggplot(<span class="pl-smi">data</span>) <span class="pl-k">+</span>
    geom_point(aes(<span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-smi">Type</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-smi">r_Adjusted</span>, <span class="pl-v">color</span><span class="pl-k">=</span><span class="pl-smi">Type</span>), <span class="pl-v">position</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>jitter<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    xlab(<span class="pl-s"><span class="pl-pds">"</span><span class="pl-pds">"</span></span>) <span class="pl-k">+</span> ylab(<span class="pl-s"><span class="pl-pds">"</span>Coevolution (Partial correlation)<span class="pl-cce">\n</span><span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    theme_bw() <span class="pl-k">+</span>
    geom_hline(<span class="pl-v">yintercept</span><span class="pl-k">=</span><span class="pl-c1">0.97</span>, <span class="pl-v">linetype</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>dashed<span class="pl-pds">"</span></span>, <span class="pl-v">alpha</span><span class="pl-k">=</span><span class="pl-c1">0.6</span>) <span class="pl-k">+</span>
    geom_hline(<span class="pl-v">yintercept</span><span class="pl-k">=</span><span class="pl-c1">0.98</span>, <span class="pl-v">linetype</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>dashed<span class="pl-pds">"</span></span>, <span class="pl-v">alpha</span><span class="pl-k">=</span><span class="pl-c1">0.6</span>) <span class="pl-k">+</span>
    geom_hline(<span class="pl-v">yintercept</span><span class="pl-k">=</span><span class="pl-c1">0.99</span>, <span class="pl-v">linetype</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>dashed<span class="pl-pds">"</span></span>, <span class="pl-v">alpha</span><span class="pl-k">=</span><span class="pl-c1">0.6</span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">0.7</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.972</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>P = 0.70<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">0.7</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.982</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>P = 0.71<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">0.7</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.992</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>P = 0.69<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">2.4</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.972</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>R = 0.72<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">2.4</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.982</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>R = 0.58<span class="pl-pds">"</span></span>) <span class="pl-k">+</span>
    annotate(<span class="pl-s"><span class="pl-pds">"</span>text<span class="pl-pds">"</span></span>, <span class="pl-v">x</span><span class="pl-k">=</span><span class="pl-c1">2.4</span>, <span class="pl-v">y</span><span class="pl-k">=</span><span class="pl-c1">0.992</span>, <span class="pl-v">label</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>R = 0.39<span class="pl-pds">"</span></span>)</pre></div>
</article></body></html>