\documentclass[11pt]{article}
\renewcommand{\baselinestretch}{1}

% PACKAGES
%-----------------------
\usepackage[left=3cm, right=3cm, top=3cm, bottom=3cm,nohead,nofoot]{geometry}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{amsmath}
\usepackage{listings}
\usepackage{graphicx}
\graphicspath{{plots/}}
\usepackage{float}
\usepackage{wrapfig}
\usepackage{caption}
\usepackage{url}
\usepackage{hyperref}
\usepackage{graphicx}
\usepackage{caption}
\usepackage{subfigure}
\usepackage[parfill]{parskip}



% DOC INFO
%-----------------------
\title{MirrorTrees: Interaction through the looking-glass}
\author{
	Sergio Castillo
	\and
	Joan Martí
	\and
	Adrià Pérez
}
\date{Structural Bioinformatics, MSc in Bioinformatics for Health Sciences - Universitat Pompeu Fabra}

% DOCUMENT
%-----------------------
\begin{document}
\maketitle

\section{MirrorTree approach}
In the so-called OMICS era, with the high-throughput techniques, the amount of data is huge. This is not an exception of proteomics data. Nowadays, huge databases for proteins exist, at distinct levels (inferred, detected, crystallised, etc). However, there is still a gap concerning the interaction between proteins. This lack of knowledge is due to the fact that proteic physic interactions used to be found by experimental means (immunoprecipitations, double-hybrid, etc).


Now, there is an opportunity to infer physical interactions \textit{in silico}, and then focus the experimental work to validation.

There are different algorithms or procedures for inferring proteic physical interactions, but in this essay we are going to cover only the MirrorTree approach, developed in early century by Dr. Alfonso Valencia and associates \cite{Pazos2001}.


MirrorTree approach is based on co-evolutionary methodology, as MirrorTree assumes that if physical interactions between proteins, they should co-evolve, moreover, they should share similar evolutionary history.



\section{Methods}
\subsection{mtree}
MirrorTree methodology, developed by Dr. Alfonso Valencia and associates, predicts direct interactions between proteins based on co-evolution of interacting proteins. \cite{Pazos2001}. The work-flow of the MirrorTree is as follows: EXPLAIN
\begin{enumerate}
\setlength{\itemsep}{1pt}
	\item Gather \textit{n} sequences to test its interaction.
	\item Look for the orthologous sequences for each query protein query (search by homology, using BLAST \cite{BLAST}).
	\item Build an MSA for each protein query.
	\item Build distance matrix for each set (using McLachlan71).
	\item Compare the the distance matrices at 1:1 rate.
	\item Build a correlation coefficient (Pearson's r) based on the difference between the matrices.
	\item Infer interaction.
\end{enumerate}
MirrorTree is settled upon the following suppositions:
\begin{enumerate}
\setlength{\itemsep}{1pt}
	\item Proteins that interact directly co-evolve.
	\item Proteins that interact directly and co-evolve should share the same evolutionary history, so they must have similar phylogenies.
\end{enumerate}
In further sections there would be a discussion over these assumptions.

\subsection{mtree}
Our integrative tool takes a slightly different approach to the Valencia's team to MirrorTrees. 

First, \texttt{mtree} performs a \texttt{jackhmmer} to search for homologous proteins for each query sequence. The program uses three iterations and then saves the resulting hidden markov models generated by \texttt{jackhmmer}. Only one homologous sequence is selected for each species, to avoid building phylogenetic trees using paralogous. The sequence with the lowest e-value for each species is used by the program.

After doing this, \texttt{mtree} considers all the possible combinations of sequences that have homologs in at least \texttt{k} common species, where \texttt{k} can be changed using the option \texttt{-sp}. The program uses a default value of \emph{10} for \texttt{k}, as similar values were used by Pazos and Valencia \cite{Pazos2001}.


In order to predict interactions, \texttt{mtree} performs an MSA for each query sequence and its homologs using \texttt{hmmalign} and the hidden markov models generated by \texttt{jackhmmer}. Then, a distance matrix is built using the BioPython module \texttt{Phylo::TreeConstructor}. Pearson's correlation coefficient and Spearman's rho is computed for each pair of matrices. 

\texttt{mtree} uses an alignment of 18S rRNA downloaded from the SILVA database \cite{SILVA} and also computes a distance matrix for each interaction candidate. Then, the program performs a partial correlation of the matrices of the two query sequences, controlling for the rRNA matrix. 




\begin{enumerate}
\item Pearson's correlation coefficient
\item Spearman's correlation coefficient
\item partial correlation 
\end{enumerate}


Our pipeline is as follows:
\begin{enumerate}
\setlength{\itemsep}{1pt}
	\item Gather \textit{n} sequences to test its interaction in .FASTA format.
	\item Download the needed databases/files:
	\begin{enumerate}
	\setlength{\itemsep}{1pt}
		\item uniprot database.
		\item Alignment of 18S rRNA sequences of the species that are used in .AFA format (already provided by animal taxon).
	\end{enumerate}
	\item Look for orthologous sequences for each query sequence (search with a hidden markov model, using jackhmmer \cite{HMMER}).
	\item Build an MSA for each query sequence and their putative orthologs (using hmmalign \cite{HMMER}).
	\item Use the distance matrix for each set provided by jackhmmer at the last iteration \cite{HMMER} (using Blosom62 matrix).
	\item Compare the the distance matrices between the query sequences.
	\item Partial out the spurious correlation (species-lineage relationship) using the distance matrix of the species tree (18S rRNA) \cite{Sato2005}.	
	\item Build a correlation coefficient (Pearson's r, Spearman's rho and partial correlations) based on the difference between the matrices.
	\item Infer interaction.
	\item Create interactome (graph).
\end{enumerate}

\subsection{Databases}
In order to develop the tool, different train/test sets where developed. Here are listed the databases from where we have retrieved the IDs or sequences: EXPLAIN
\begin{enumerate}
\setlength{\itemsep}{1pt}
	\item Sequences in FASTA format (Uniprot-Swiss database).
	\item ID's from proteins that interact (IntAct database \cite{intact}).
	\item ID's from proteins that do not interact (Negatome database \cite{negatome}).
	\item Alignments of 18S rRNA for the species tree (SILVA format \cite{SILVA}).
\end{enumerate}

\section{Discussion}
Despite being a possible theoretical approach, Valencia's work in 2001\cite{Pazos2001} needs to be improved and analysed. The approach here presented has prominent differences with the reference work. In the following sections there are specified which points needed to be changed and the reasons for doing so.

\subsection{Assumptions}
One of the main handicaps that exist in MirrorTree approaching is that its assumptions are not always valid.
\begin{enumerate}
\setlength{\itemsep}{1pt}
	\item \textit{Proteins that interact directly co-evolve}: Not necessarily. Interaction between proteins 		occurs at domain level, more precisely it interacts with the interaction region, a smaller region within 		the domain, not at the whole protein. It might be true that interaction regions co-evolve, but that does 		not mean that co-evolve the whole protein. Besides, even the co-evolution between interaction sites is 		somehow obscure, as it depends on the intensity of the selective pressure between them. It is known that 		the interaction between proteins is not static, is partially adaptative (key-lock vs greeting-hands). So, 	would always be relevant an aminoacid exchange? How intense is the selective pressure between the 			regions?

	Moreover, co-evolution can also act at network level, therefore, it is plausible that proteins that do 		not physically interact may co-evolve, as both might be of importance in the pathway. Then, does proteins 	can de detected as false positive, as they interact, they have a relationship but not interact directly. 		Co-evolution is a good measure for direct interaction?

	\item \textit{Proteins that interact directly and co-evolve should share the same evolutionary history, 		so they may have similar phylogenies.} This is a strong assumption, as it would be commented in following		sections, co-evolution relies upon strong selective pressure in order to be fully detected. In order to 		have coincident phylogenies, there should be a fairly strong selection between interactors. The problem 		raises as this is not a frequent.
\end{enumerate}

\subsection{Search of orthologs}
One of the main difficulties when working with phylogenies, and by extension to this essay, is the difficulty to distinguish orthologs from paralogs, one of the paradigmatic problems on the filed. 

When retrieving orthologous sequences, paralogous sequences are also retrieved. As they refer to different events of diversification (species vs. gene respectively), there is an introgression of an other relationship within our model, adding confounding relationships, and therefore decreasing the predictive power of the approach.

Besides, proteins, and in special, proteins from multicellular eukaryots, may have many domains. If the aim of the study is to asses direct interaction between proteins, it must focus on the interaction region. Using BLAST-based searches can also lead to misleading homology, as it might search homologous sequences based on a domain which is not related to the interaction.

On the contrary, HMM-based searches look for homologous sequences using an specific domain, which is fetched from the query sequence. Doing so, relationship between proteins, focusing on the interacting domain, can be ensured.


\subsection{Weight Matrix}
The distance matrix can be computed using different weight matrices (p.e. Blosum62, McLachlan71, Dayhoff Pam Matrix, etc) or distance correctors (p.e. Kimura's distance). Despite being a variable in the approach, some authors\cite{Zhou13} have found that its change does not provide an improvement of the prediction power.

This might be true due to that the focus of MirrorTree is to compare between trees (a.k.a distance matrix). The relevant point is to use the same statistic in both MSAs, not in how good or accurate is the estimation of the distances.

\subsection{Correlation measures}
Searching on the main bibliography of the topic, 3 measures of correlation are used:

\begin{enumerate}
\setlength{\itemsep}{1pt}
	\item \textit{Pearson's correlation coefficient}\cite{Pazos2001}.
	\item \textit{Spearman's rho:} As maybe the relationship between matrices is not linear, or it may have 		outlayers that alter the fitting of the linear model, it could be of help to see how it behaves a rank-		based correlation.
	\item \textit{Partial correlation (using 18S rRNA phylogeny of the species):}\cite{Sato2005} When 			performing correlation between matrices, exist a common relationship that need to be taken into account, 		the phylogeny of the species used in the trees. For building the phylogeny a lineage-specific marker is 		needed. Usually is recommended to use the 18S rRNA for these purpose\cite{Sato2005}. It would be 				interesting to see how the correlation changes.
\end{enumerate}

In order to see which of them provided a better classification for interactions between proteins, all three statistics where build.

\subsection{Reproducibility}
One of the reported problems about the MirrorTree approach is the reproducibility. Above all, the goodness of the results depend on the chosen dataset. As observed in Zhou's work\cite{Zhou13}, 3 different sets, with reasonable resemblance between them:
\begin{enumerate}
\setlength{\itemsep}{1pt}
	\item \textit{Set1:} Common interactors of \textit{Saccharomyces cerevisiae} and \textit{Homo sapiens}, 		with both cases experimentally confirmed their interaction.
	\item \textit{Set2}: Analogous but pairs only need to be experimentally confirmed in one of the two 			species.
	\item \textit{Set3}: Analogous to set1 but using common interactors of \textit{Mus musculus} and   \textit{Homo sapiens}).
\end{enumerate}
Only certain degree of differences between interacting pairs and non-interacting pairs can be seen in set1. Set2, a more stringent subset of set1, does not show any difference between both groups, and set3, which is analogous to set1 but using more evolutionary closed species, neither show it.

In addition, in Valencia's article\cite{Pazos2001} 3 datasets are used:
\begin{enumerate}
\setlength{\itemsep}{1pt}
	\item \textit{Set1:} Interacting Domains.
	\item \textit{Set2:} Interacting Proteins of bacteria.
	\item \textit{Set2:} Interacting Proteins by genome.
\end{enumerate}
The results of the set1 (domains) is highly controversial, as it predicts better the co-evolution between domains of the same protein (logic if within protein selective pressure is taken into account) than between interacting domains. As previously mentioned, maybe the selective pressure, and thus the co-evolution is laxer in interacting domains than in within protein domains. Therefore, is co-evolution a good measure of selective pressure in protein interaction?
In set2, it seems to be possible to properly classify the interactions (despite counting with ambiguous classification \textit{possible}). In set 3 the situation is analogous to set2.


\section{Results}
\subsection{Mutual information}
\subsection{Data}
\includegraphics{partial_corr}


\bibliographystyle{plain}
\bibliography{project}	
\end{document}
